File: jbake-core/src/main/java/org/jbake/util/ConfigurationPrinter.java
package org.jbake.util;

import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.app.configuration.Property;

import java.io.PrintStream;
import java.util.List;

public class ConfigurationPrinter {

    private PrintStream out;
    private JBakeConfiguration configuration;

    public ConfigurationPrinter(JBakeConfiguration configuration, PrintStream out) {
        this.out = out;
        this.configuration = configuration;
    }

    public void print() {

        List<Property> properties = configuration.getJbakeProperties();
        Property.Group lastGroup = null;

        for (Property property : properties) {

            if (lastGroup != property.getGroup()) {
                printGroup(property);
                this.printHeader();
            }
            if (!property.getDescription().isEmpty()) {
                printDescription(property);
            }
            printKeyAndValue(property);

            lastGroup = property.getGroup();
        }
    }

    private void printHeader() {
        out.printf("%1$-40s %2$-40s%n", "Key", "Value");
        out.println(getHorizontalLine());
    }

    private String getHorizontalLine() {
        return String.format("%080d%n", 0).replace("0", "-");
    }

    private void printGroup(Property property) {
        out.printf("%n%s - Settings%n%n", property.getGroup());
    }

    private void printDescription(Property property) {
        out.printf("# %s%n", property.getDescription());
    }

    private void printKeyAndValue(Property property) {
        String key = leftFillWithDots(property.getKey());
        Object value = configuration.get(property.getKey());
        out.printf("%1$s: %2$-40s%n%n", key, value);
    }

    private String leftFillWithDots(String value) {
        return String.format("%1$-40s", value).replace(' ', '.');
    }

}

File: jbake-core/src/main/kotlin/org/jbake/util/ConfigurationPrinter.kt
package org.jbake.util

import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.app.configuration.JBakeProperty
import java.io.PrintStream

class ConfigurationPrinter(private val configuration: JBakeConfiguration, private val out: PrintStream) {
    fun print() {
        val properties: MutableList<JBakeProperty> = configuration.jbakeProperties
        var lastGroup: JBakeProperty.Group? = null

        for (property in properties) {
            if (lastGroup != property.group) {
                printGroup(property)
                this.printHeader()
            }
            if (property.description.isNotEmpty()) printDescription(property)
            printKeyAndValue(property)

            lastGroup = property.group
        }
    }

    private fun printHeader() {
        out.printf("%1$-40s %2$-40s%n", "Key", "Value")
        out.println(this.horizontalLine)
    }

    private val horizontalLine by lazy { String.format("%080d%n", 0).replace("0", "-") }

    private fun printGroup(property: JBakeProperty) = out.printf("%n%s - Settings%n%n", property.group)

    private fun printDescription(property: JBakeProperty) = out.printf("# %s%n", property.description)

    private fun printKeyAndValue(property: JBakeProperty) {
        val key = leftFillWithDots(property.key)
        val value = configuration.get(property.key)
        out.printf("%1\$s: %2\$-40s%n%n", key, value)
    }

    private fun leftFillWithDots(value: String): String {
        return String.format("%1$-40s", value).replace(' ', '.')
    }
}
